src/microbe_stage/CompoundCloudSystem.cs

```diff
@@ -213,134 +185,167 @@ public float AmountAvailable(Compound compound, Vector3 worldPosition, float fra
             throw new ArgumentException("searchRadius must be >= 1");
 
         int resolution = Resolution;
-
-        // This version is used when working with cloud local coordinates
         float localRadius = searchRadius / resolution;
-
+        float localRadiusSquared = localRadius * localRadius;
         float nearestDistanceSquared = float.MaxValue;
-
         Vector3? closestPoint = null;
+        int compoundID = GetCompoundID(compound);
 
-        foreach (var cloud in clouds)
+        foreach (var cloudInstance in cloudInstances)
         {
-            // Skip clouds that don't handle the target compound
-            if (!cloud.HandlesCompound(compound))
+            var compounds = (int[])cloudInstance.Call("get_compounds");
+
+            if (!Array.Exists(compounds, id => id == compoundID))
                 continue;
 
-            // Skip clouds that are out of range
-            if (!cloud.ContainsPositionWithRadius(position, searchRadius))
+            bool contains = (bool)cloudInstance.Call("contains_position_with_radius", position, searchRadius);
+            if (!contains)
                 continue;
 
-            cloud.ConvertToCloudLocal(position, out var cloudRelativeX, out var cloudRelativeY);
+            var coords = (Vector2)cloudInstance.Call("convert_to_cloud_local", position);
+            int cloudRelativeX = (int)coords.X;
+            int cloudRelativeY = (int)coords.Y;
+
+            int cloudSize = Size;
+
+            int xStart = Math.Max(0, cloudRelativeX - (int)localRadius);
+            int xEnd = Math.Min(cloudSize - 1, cloudRelativeX + (int)localRadius);
+            int yStart = Math.Max(0, cloudRelativeY - (int)localRadius);
+            int yEnd = Math.Min(cloudSize - 1, cloudRelativeY + (int)localRadius);
 
-            // Search each angle for nearby compounds
-            for (int radius = 1; radius < localRadius; radius += 1)
+            for (int x = xStart; x <= xEnd; x++)
             {
-                for (double theta = 0; theta <= MathUtils.FULL_CIRCLE; theta += Constants.CHEMORECEPTOR_ARC_SIZE)
+                for (int y = yStart; y <= yEnd; y++)
                 {
-                    int x = cloudRelativeX + (int)Math.Round(Math.Cos(theta) * radius);
-                    int y = cloudRelativeY + (int)Math.Round(Math.Sin(theta) * radius);
+                    float dx = x - cloudRelativeX;
+                    float dy = y - cloudRelativeY;
+                    float distanceSquared = dx * dx + dy * dy;
 
-                    // Negative coordinates are always outside the cloud area
-                    if (x < 0 || y < 0)
+                    if (distanceSquared > localRadiusSquared)
                         continue;
 
-                    // Then just need to check that it is within the cloud simulation array
-                    if (x < cloud.Size && y < cloud.Size)
+                    float amount = (float)cloudInstance.Call("amount_available", compoundID, x, y, 1.0f);
```

Comment:
This is in a very big loop calling a ton of methods through the Godot calling interface.